### PR TITLE
Dynamic control node receptor config

### PR DESCRIPTION
**Pull request by TheRealHaoLiu**
##### SUMMARY
when deploy AWX controlplane is deployed on Kubernetes, generate the receptor.conf that peer out to all external execution instance

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME
 - API

##### AWX VERSION
```
awx: 21.3.1.dev57+g2fbedfb47b
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Review by fosterseth — APPROVED** [review]

File: tools/docker-compose/ansible/roles/sources/templates/docker-compose.yml.j2 [hunk 1]
```diff
@@ -39,6 +39,7 @@ services:
       - "../../docker-compose/_sources/local_settings.py:/etc/tower/conf.d/local_settings.py"
       - "../../docker-compose/_sources/SECRET_KEY:/etc/tower/SECRET_KEY"
       - "../../docker-compose/_sources/receptor/receptor-awx-{{ loop.index }}.conf:/etc/receptor/receptor.conf"
+      - "../../docker-compose/_sources/receptor/receptor-awx-{{ loop.index }}.conf.lock:/etc/receptor/receptor.conf.lock"
```
Comment (shanemcd) [inline tools/docker-compose/ansible/roles/sources/templates/docker-compose.yml.j2:42]: This feels wrong to me. Give me until tomorrow to play around with another option.

File: awx/main/tasks/receptor.py [hunk 2]
```diff
@@ -570,3 +592,66 @@ def kube_config(self):
...
+)
+
+
+@task()
+def write_receptor_config():
+    receptor_config = list(RECEPTOR_CONFIG_STARTER)
+
+    instances = Instance.objects.exclude(node_type='control')
+    for instance in instances:
+        peer = {'tcp-peer': {'address': f'{instance.hostname}:{instance.listener_port}', 'tls': 'tlsclient'}}
```
Comment (shanemcd) [inline awx/main/tasks/receptor.py:636]: 👍

File: tools/docker-compose/ansible/roles/sources/templates/docker-compose.yml.j2 (hunk 1 above)
Comment (TheRealHaoLiu) [reply tools/docker-compose/ansible/roles/sources/templates/docker-compose.yml.j2:42]: yah this feels wrong...if i want to fix it i would create multiple dirs under _source and each will be for a different receptor node this way i can just mount the whole dir

File: awx/main/models/ha.py [hunk 3]
```diff
@@ -419,6 +419,12 @@ def on_instance_group_saved(sender, instance, created=False, raw=False, **kwargs
 
 @receiver(post_save, sender=Instance)
 def on_instance_saved(sender, instance, created=False, raw=False, **kwargs):
+    # TODO: this needs to be smarter. Did the port change, etc?
+    if settings.IS_K8S and created and instance.node_type == 'execution':
```
Comment (jbradberry) [inline awx/main/models/ha.py]: What about hop nodes? Last I knew, hop nodes are 'in' for this release.

File: awx/main/models/ha.py [hunk 4]
```diff
@@ -419,6 +419,12 @@ def on_instance_group_saved(sender, instance, created=False, raw=False, **kwargs
...
+        from awx.main.tasks.receptor import write_receptor_config  # prevents circular import
+
+        connection.on_commit(lambda: write_receptor_config.apply_async(queue='tower_broadcast_all'))
```
Comment (jbradberry) [inline awx/main/models/ha.py:431]: I'd kind of like a comment here.  I totally missed the fact that the tasks were going to be fired off across all instances in the control-plane.

File: awx/main/tasks/jobs.py [hunk 5]
```diff
@@ -146,7 +146,7 @@ def build_execution_environment_params(self, instance, private_data_dir):
         """
         Return params structure to be executed by the container runtime
         """
-        if settings.IS_K8S:
+        if settings.IS_K8S and instance.instance_group.is_container_group:
```
Comment (AlanCoding) [inline awx/main/tasks/jobs.py:149]: :thinking: what about project updates? Their instance group isn't a container group, so it should work by that metric. I would like this `is_container_group` value to be more reliable, or maybe be obvious on `instance` itself.

Anyway, this should work but we need to see it in action with an execution node at some point.

File: awx/main/tasks/receptor.py [hunk 6]
```diff
@@ -25,12 +25,17 @@
     cleanup_new_process,
 )
 from awx.main.constants import MAX_ISOLATED_PATH_COLON_DELIMITER
+from awx.main.models import Instance
+from awx.main.dispatch.publish import task
 
 # Receptorctl
 from receptorctl.socket_interface import ReceptorControl
 
+from filelock import FileLock
```
Comment (AlanCoding) [inline awx/main/tasks/receptor.py:34]: maybe I'm dense, but I can't figure out where this requirement is coming from.

Comment (TheRealHaoLiu) [reply awx/main/tasks/receptor.py:34]: can u elaborate on this comment, are you asking about "why do we need to have file lock here?"

Comment (AlanCoding) [reply awx/main/tasks/receptor.py:34]: We could probably use this in `acquire_lock` if it's really there.

Comment (AlanCoding) [reply awx/main/tasks/receptor.py:34]: No, I mean, wouldn't you need to add this to the `requirements.txt`? Or is it in the standard library? Testing on python 3.10 I don't see it.

File: awx/main/models/ha.py (hunk 4 above)
Comment (TheRealHaoLiu) [reply awx/main/models/ha.py:431]: done

File: awx/main/tasks/receptor.py (hunk 2 above)
Comment (TheRealHaoLiu) [reply awx/main/tasks/receptor.py:636]: 👍

File: tools/docker-compose/ansible/roles/sources/templates/docker-compose.yml.j2 (hunk 1 above)
Comment (TheRealHaoLiu) [reply tools/docker-compose/ansible/roles/sources/templates/docker-compose.yml.j2:42]: we can address it in follow up PR

File: awx/main/tasks/receptor.py (hunk 6 above)
Comment (TheRealHaoLiu) [reply awx/main/tasks/receptor.py:34]: weird... i try to run this in the development environment and it works 

well i haven't build a "normal" image yet. ur right filelock is not a standard library i added to requirements/requirements.in

File: requirements/requirements.in [hunk 7]
```diff
@@ -25,6 +25,7 @@ django-split-settings
 django-taggit
 djangorestframework==3.13.1
 djangorestframework-yaml
+filelock
```
Comment (jbradberry) [inline requirements/requirements.in:28]: You've got to do the pip-compile dance.  https://github.com/ansible/awx/blob/devel/requirements/README.md#upgrading-or-adding-select-libraries

File: awx/main/models/ha.py [hunk 8]
```diff
@@ -423,6 +423,13 @@ def on_instance_group_saved(sender, instance, created=False, raw=False, **kwargs
 
 @receiver(post_save, sender=Instance)
 def on_instance_saved(sender, instance, created=False, raw=False, **kwargs):
+    # TODO: this needs to be smarter. Did the port change, etc?
+    if settings.IS_K8S and created and (instance.node_type == 'execution' or instance.node_type == 'hop'):
```
Comment (jbradberry) [inline awx/main/models/ha.py]: Optional picky: it would have been more compact to use `instance.node_type in (...)`

**Review by jbradberry — CHANGES_REQUESTED** [review]

File: requirements/requirements.in (hunk 7 above)
Comment (TheRealHaoLiu) [reply requirements/requirements.in:28]: ugg... done

**Review by jbradberry — APPROVED** [review]

**Pull request merged**